doc/developer/design/20220214_create_cluster.md

```diff
@@ -0,0 +1,191 @@
+
+The syntax of `SHOW INDEXES` will be extended to allow limiting the indexes
+shown to one cluster:
+
+```
+SHOW { INDEX | INDEXES | KEYS } { FROM | IN } <on_name>
+[IN CLUSTER <cluster_name>]
+{ LIKE 'pattern | EXPR <expr> }
+```
+
+The output of `SHOW INDEXES` will be extended with an additional `cluster_name`
+column which displays the name of the cluster in which the index is
+materialized.
+
+### SQL session changes
+
+A new `cluster` session variable will be introduced that specifies the "active"
+cluster for a SQL session. It is analogous to the existing `database` session
+variable.
+
+The default value for the `cluster` session variable will be initially hardcoded
+to `default`. Users can change the value on a per-session basis via the usual
+mechanisms, like the pgwire startup protocol and the `SET` command.
+
+See [future work](#future-work) for details on how we plan to make the default
+configurable in the future.
+
+`SELECT` statements issued in a session will only use the indexes from the
+active cluster.
+
+### System catalog changes
+
+A new `mz_clusters` table with the following structure will describe the
+available clusters in the system:
+
+Field   | Type      | Meaning
+--------|-----------|--------
+id      | `bigint`  | The ID of the cluster.
+name    | `text`    | The name of the cluster.
+virtual | `boolean` | Whether the cluster is virtual.
+size    | `text`    | The size of the cluster.
+
+The existing `mz_indexes` table will be extended with a `cluster_id` field:
+
+Field      | Type      | Meaning
+-----------|-----------|--------
+...        | ...       | ...
+cluster_id | `bigint`  | The ID of the cluster in which the index is materialized.
+
+### Coordinator changes
+
+Internally, each cluster will be identified by an identifier in the range `[1,
+i64::MAX]`. The coordinator will be responsible for allocating identifiers and
+durably recording the mapping from names to identifiers via the following new
+table in the SQLite catalog:
+
+```sql
+CREATE TABLE clusters (
+    id   integer PRIMARY KEY,
+    name text NOT NULL UNIQUE
+);
+```
+
+This table works analogously to the existing `databases` table in the catalog.
+
+The migration that creates the table will additionally install a default row
+of `(1, 'default')`. This creates a cluster named `default` that is present by
+default in every Materialize installation. The default cluster is not otherwise
+special and can be dropped if desired.
+
+We will write a catalog migration that adds the `IN CLUSTER default` clause to
+all existing indexes.
+
+The various bookkeeping in the coordinator will need to change to account for
+clusters. Execution of a `SELECT` statement, for example, will need to select
+from only the indexes in the active cluster.
+
+TODO(mcsherry?): can we more specifically describe what bookkeeping will change?
+
+### `mz_cluster_id`
+
+The existing `mz_cluster_id` function uses "cluster" to mean something entirely
```

Comment:
Do we need to keep this thing? We removed mz_workers wholesale and hard coded it into the telemetry stats. If we don't need this function, we could remove it instead of giving it some name with `legacy` in it. If we do need it, it should perhaps not be labeled as "legacy"?

Reply:
> We removed mz_workers wholesale and hard coded it into the telemetry stats.

I think we can actually `mz_workers` back, now that we have clusters. It just needs to read the number of workers from `mz_clusters`, once that's exposed. I guess it also needs to wait on the nonpure function PR to land!

I'm not sure whether we want to keep the existing telemetry infrastructure or not. If we do, we probably want to track things by account IDs, and invent account IDs for on-premise deployments. I basically just want to kick the can down the road on that part of the design. Maybe we can just move `mz_cluster_id` into `mz_internal` and leave its name alone for now.